test/unit/org/apache/cassandra/index/sai/utils/RangeIntersectionIteratorTest.java

```diff
@@ -0,0 +1,435 @@
+        assertEquals(6L, builder.getMaximum().token().getLongValue());
+        assertEquals(9L, builder.getTokenCount());
+        assertEquals(3L, builder.rangeCount());
+        Assert.assertFalse(builder.statistics.isDisjoint());
+
+        assertEquals(1L, builder.rangeIterators.get(0).getMinimum().token().getLongValue());
+        assertEquals(4L, builder.rangeIterators.get(1).getMinimum().token().getLongValue());
+        assertEquals(6L, builder.rangeIterators.get(2).getMinimum().token().getLongValue());
+
+        builder.add(new LongIterator(new long[] { 1L, 2L, 6L }));
+        builder.add(new LongIterator(new long[] { 4L, 5L, 6L }));
+        builder.add(new LongIterator(new long[] { 6L, 8L, 9L }));
+
+        assertEquals(convert(6L), convert(builder.build()));
+
+        builder = RangeIntersectionIterator.builder();
+        builder.add(new LongIterator(new long[]{ 1L, 5L, 6L }));
+        builder.add(new LongIterator(new long[]{ 3L, 5L, 6L }));
+
+        RangeIterator tokens = builder.build();
+
+        assertEquals(convert(5L, 6L), convert(tokens));
+
+        FileUtils.closeQuietly(tokens);
+
+        RangeIterator emptyTokens = RangeIntersectionIterator.builder().build();
+        assertEquals(0, emptyTokens.getCount());
+
+        builder = RangeIntersectionIterator.builder();
+        assertEquals(0L, builder.add((RangeIterator) null).rangeCount());
+        assertEquals(0L, builder.add((List<RangeIterator>) null).getTokenCount());
+        assertEquals(0L, builder.add(new LongIterator(new long[] {})).rangeCount());
+
+        RangeIterator single = new LongIterator(new long[] { 1L, 2L, 3L });
+        RangeIterator range = RangeIntersectionIterator.builder().add(single).build();
+
+        // because build should return first element if it's only one instead of building yet another iterator
+        assertEquals(range, single);
+
+        // Make a difference between empty and null ranges.
+        builder = RangeIntersectionIterator.builder();
+        builder.add(new LongIterator(new long[] {}));
+        assertEquals(0L, builder.rangeCount());
+        builder.add(single);
+        assertEquals(1L, builder.rangeCount());
+        range = builder.build();
+        assertEquals(0, range.getCount());
+
+        // disjoint case
+        builder = RangeIntersectionIterator.builder();
+        builder.add(new LongIterator(new long[] { 1L, 2L, 3L }));
+        builder.add(new LongIterator(new long[] { 4L, 5L, 6L }));
+
+        Assert.assertTrue(builder.statistics.isDisjoint());
+
+        RangeIterator disjointIntersection = builder.build();
+        Assert.assertNotNull(disjointIntersection);
+        Assert.assertFalse(disjointIntersection.hasNext());
+
+    }
+
+    @Test
+    public void emptyRangeTest()
+    {
+        RangeIterator.Builder builder;
+
+        // empty, then non-empty
+        builder = RangeIntersectionIterator.builder();
+        builder.add(new LongIterator(new long[] {}));
```

Comment:
nit: What if we just threw in a constructor for `LongIterator` that takes no arguments and calls `super(null, null, 0);` ?

Reply:
...or a new static factory like `newEmptyIterator()`